docs/core/v3.0.rst

```diff
@@ -0,0 +1,1762 @@
+    interface`_ which is a common set of operations that stores may
+    provide.
+
+.. _storage transformer:
+.. _storage transformers:
+
+*Storage transformer*
+
+    To enhance the storage capabilities, storage transformers may
+    change the storage structure and behaviour of data coming from
+    an array_ in the underlying store_. Upon retrieval the original data
+    is restored within the transformer. Any number of `predefined storage
+    transformers`_ can be registered and stacked.
+    See the `storage transformers details`_ below.
+
+.. _`storage transformers details`: #storage-transformers-1
+
+Node names
+==========
+
+Except for the root node, each node in a hierarchy must have a name,
+which is a string of characters. To ensure consistent behaviour
+across different storage systems, the following constraints apply to
+node names:
+
+* must not be the empty string ("")
+
+* must use only characters in the sets ``a-z``, ``A-Z``, ``0-9``,
+  ``-_.``
+
+* must not be a string composed only of period characters, e.g. "." or
+  ".."
+
+* must be at most 255 characters long
+
+Node names are case sensitive, e.g., the names "foo" and "FOO" are **not**
+identical.
+
+.. note:
+    The Zarr core development team recognises that restricting the set
+    of allowed characters creates an impediment and bias against users
+    of different languages. We are actively discussing whether the full
+    Unicode character set could be allowed and what technical issues
+    this would entail. If you have experience or views please comment on
+    `issue #56 <https://github.com/zarr-developers/zarr-specs/issues/56>`_.
+
+
+Data types
+==========
+
+A data type describes the set of possible binary values that an array
+element may take, along with some information about how the values
+should be interpreted.
+
+This core specification defines a limited set of data types to
+represent boolean values, integers, and floating point
+numbers. Extensions may define additional data types. All of the data
+types defined here have a fixed size, in the sense that all values
+require the same number of bytes. However, extensions may define
+variable sized data types.
+
+Note that the Zarr specification is intended to enable communication
+of data between a variety of computing environments. The native byte
+order may differ between machines used to write and read the data.
+
+Each data type is associated with an identifier, which can be used in
+metadata documents to refer to the data type. For the data types
+defined in this specification, the identifier is a simple ASCII
+string. However, extensions may use any JSON value to identify a data
+type.
+
+
+Core data types
+---------------
+
+.. list-table:: Data types
+   :header-rows: 1
+
+   * - Identifier
+     - Numerical type
+     - Size (no. bytes)
+     - Byte order
+   * - ``bool``
+     - Boolean, with False encoded as ``\\x00`` and True encoded as ``\\x01``
+     - 1
+     - None
```

Comment:
I think it would be better to choose a different naming scheme for data types that allows for more extensibility without introducing inconsistency.  In particular, "<f2" is already taken to mean float16, so there is no natural choice for bfloat16.

Instead I would propose "bool", "int8", "uint8", "float16", "float32", etc. as the names, and leave endianness as part of the filter/codec.  But if endianness is kept as part of the data type, then "uint16le", "float32le", etc.  See also: https://github.com/zarr-developers/zarr-specs/issues/130 and https://github.com/zarr-developers/zarr-specs/issues/131

Reply:
I hadn't considered the endianness as a filter/codec but I really like this idea. TypedArrays (the container for array data) in JavaScript don't handle endianess, so in zarr.js we essentially support be/le with a custom filter (which flips bytes if necessary to the user system).

Reply:
I agree with @jbms. We probably should denote types explicitly in terms of bits, rather than bytes, if possible. The C23 standardization is considering incorporating `BitInt` for arbitrary bit sized integers, which may be important for Zarr in the future.

https://reviews.llvm.org/D108643

`i8` also means a 8-bit integer rather than a 64-bit integer in Rust and LLVM IR. The numpy short form notation is confusing in a polyglot context.

https://doc.rust-lang.org/book/ch03-02-data-types.html
https://llvm.org/docs/LangRef.html#integer-type

Reply:
Fully agree here, the numpy dtype naming convention is unfortunate and confusing in other languages, also c++, where you would write everything in bits (e.g. `int8` etc.)

Reply:
The bit/byte naming inconsistency was also one of our few notes on the spec from our v3 implementation. Also, while niche but useful types like bit-packed uint10 are unlikely to ever be in any core spec, the more verbose  scheme allows extensions to name their types more consistently with the core set.

Reply:
Resolved via #155